packages/tre/src/plugins/cache/gateway.ts

```diff
@@ -130,137 +142,186 @@ function getMatchResponse(
-      status: this.metadata.status,
-      headers: this.metadata.headers,
-    });
-  }
-}
-
-interface ParsedHttpResponse {
-  headers: Headers;
-  status: number;
-  body: Uint8Array;
-}
 class HttpParser {
-  readonly server: http.Server;
-  readonly responses: Map<string, Uint8Array> = new Map();
-  readonly connected: Promise<void>;
   private static INSTANCE: HttpParser;
   static get(): HttpParser {
     HttpParser.INSTANCE ??= new HttpParser();
     return HttpParser.INSTANCE;
   }
+
+  readonly #responses: Map<string, ReadableStream<Uint8Array>> = new Map();
+  readonly #ready: Promise<URL>;
+
   private constructor() {
-    this.server = http.createServer(this.listen.bind(this)).unref();
-    this.connected = new Promise((accept) => {
-      this.server.listen(0, "localhost", accept);
+    const server = http.createServer(this.#listen).unref();
+    this.#ready = new Promise((resolve) => {
+      server.listen(0, "localhost", () => {
+        const address = server.address();
+        assert(address !== null && typeof address === "object");
+        resolve(new URL(`http://localhost:${address.port}`));
+      });
     });
   }
-  private listen(request: http.IncomingMessage, response: http.ServerResponse) {
-    assert(request.url !== undefined);
-    assert(response.socket !== null);
-    const array = this.responses.get(request.url);
-    assert(array !== undefined);
+
+  #listen: http.RequestListener = async (req, res) => {
+    assert(req.url !== undefined);
+    assert(res.socket !== null);
+    const stream = this.#responses.get(req.url);
+    assert(stream !== undefined);
     // Write response to parse directly to underlying socket
-    response.socket.write(array);
-    response.socket.end();
-  }
-  public async parse(response: Uint8Array): Promise<ParsedHttpResponse> {
-    await this.connected;
+    for await (const chunk of stream) res.socket.write(chunk);
+    res.socket.end();
+  };
+
+  async parse(response: ReadableStream<Uint8Array>): Promise<Response> {
+    const baseURL = await this.#ready;
     // Since multiple parses can be in-flight at once, an identifier is needed
     const id = `/${crypto.randomBytes(16).toString("hex")}`;
-    this.responses.set(id, response);
-    const address = this.server.address()! as AddressInfo;
+    this.#responses.set(id, response);
     try {
-      const parsedResponse = await fetch(
-        `http://localhost:${address.port}${id}`
-      );
-      const body = await parsedResponse.arrayBuffer();
-      return {
-        headers: parsedResponse.headers,
-        status: parsedResponse.status,
-        body: new Uint8Array(body),
-      };
+      return await fetch(new URL(id, baseURL));
     } finally {
-      this.responses.delete(id);
+      this.#responses.delete(id);
     }
   }
 }
 
+class SizingStream extends TransformStream<Uint8Array, Uint8Array> {
+  readonly size: Promise<number>;
+
+  constructor() {
+    const sizePromise = new DeferredPromise<number>();
```

Comment:
I'm guessing the Stream Promises API doesn't work with web streams> https://nodejs.org/api/stream.html#streams-promises-api